lib/Sema/TypeCheckNameLookup.cpp

```diff
@@ -214,6 +214,43 @@ TypeChecker::lookupUnqualifiedType(DeclContext *dc, DeclName name,
                                    NameLookupOptions options) {
   SmallVector<TypeDecl *, 1> decls;
 
+  // CodingKeys is a special type. Although not a protocol requirement of the
+  // Encodable/Decodable protocols, it is synthesized as part of
+  // Encodable/Decodable conformance. Unqualified lookup may occur before the
+  // type is synthesized, though, in which case, we need to make sure it gets
+  // synthesized.
+  //
+  // If this is what we're looking for, then we need to attempt to cause the
+  // type to be synthesized before performing the lookup.
+  if (name.isSimpleName(Context.Id_CodingKeys)) {
```

Comment:
I'd love to do it as part of the lookup below but unfortunately, `UnqualifiedLookup` has no knowledge of `TypeChecker` (`TypeChecker.h` isn't even exposed in `include/swift`; it's private to `lib/Sema`), so we can't do things like checking types for protocol conformance and synthesis. Although it makes more sense in there, I felt like it was a more sweeping change to tightly couple things in `lib/AST` with things in `lib/Sema`.